mlir/tools/rocmlir-tuning-driver/rocmlir-tuning-driver.cpp

```diff
@@ -131,79 +126,101 @@ static FailureOr<double> benchmarkKernel(const char *binary,
+extractKernels(ModuleOp op, SmallVector<func::FuncOp> &kernels) {
   if (!op->hasAttr("mhal.arch")) {
     return op->emitOpError(
         "no architecture set, set mhal.arch on the input module");
   }
-  rock::RockGemmWrapperInterface kernel;
-  uint32_t nKernels = 0;
-  op.walk([&kernel, &nKernels](rock::RockGemmWrapperInterface candidate) {
-    nKernels++;
-    kernel = candidate;
+  rock::RockGemmWrapperInterface toTune;
+  op.walk([&toTune, &kernels](func::FuncOp f) {
+    Attribute kernel = f->getAttr("kernel");
+    if (!kernel)
+      return;
+    kernels.push_back(f);
+
+    f.walk([&toTune](rock::RockGemmWrapperInterface gemmLike) {
+      if (!toTune)
+        toTune = gemmLike;
+    });
   });
-  if (nKernels == 0)
-    return op.emitOpError("input module contains no kernels");
-  if (nKernels > 1)
-    return op.emitOpError(
-        "more than one kernel on the input, don't know what to tune");
-  return kernel;
+
+  std::sort(kernels.begin(), kernels.end(),
+            [](const func::FuncOp &a, const func::FuncOp &b) {
+              int kernelA = a->getAttrOfType<IntegerAttr>("kernel").getInt();
+              int kernelB = b->getAttrOfType<IntegerAttr>("kernel").getInt();
+              return kernelA < kernelB;
+            });
+
+  return toTune;
 }
 
 static LogicalResult runTuningLoop(ModuleOp source) {
   // Verify prerequisites
-  FailureOr<rock::RockGemmWrapperInterface> maybeToTune = extractKernel(source);
+  SmallVector<func::FuncOp> funcs;
+  FailureOr<rock::RockGemmWrapperInterface> maybeToTune =
+      extractKernels(source, funcs);
   if (failed(maybeToTune))
     return failure();
   rock::RockGemmWrapperInterface toTune = std::move(*maybeToTune);
   // Provisionally use the type of input A to set up the init value - this
   // should be a per-buffer value in the futurue.
   benchmark::DataType dataType = getDataType(toTune.getAType());
 
-  auto kernelFunc = toTune->getParentOfType<func::FuncOp>();
-  if (!kernelFunc || !kernelFunc->hasAttr("kernel"))
-    return toTune.emitOpError(
-        "kernel must be in a function with the kernel attribute");
-
   // We need a copy since HIP'll want a C string
-  std::string kernelFuncName = kernelFunc.getSymName().str();
-  std::vector<size_t> bufferLengths;
-  for (Type argType : kernelFunc.getArgumentTypes()) {
-    auto shapedTy = argType.dyn_cast<ShapedType>();
-    if (!shapedTy)
-      return kernelFunc.emitOpError("all kernel inputs must be shaped types");
-    if (!shapedTy.hasStaticShape())
-      return kernelFunc.emitOpError(
-          "all kernel arguments must have static shape");
-    int64_t sizeInBits =
-        shapedTy.getNumElements() * shapedTy.getElementTypeBitWidth();
-    bufferLengths.push_back(sizeInBits / 8);
+  SmallVector<std::string> kernelFuncNames;
+  SmallVector<size_t> bufferLengths;
+  for (func::FuncOp &funcOp : funcs) {
+    kernelFuncNames.push_back(funcOp.getSymName().str());
+    for (Type argType : funcOp.getArgumentTypes()) {
```

Comment:
This'll push the sizes multiple times. You'll want to run this code on `funcs[0]`